### ISSUE TITLE
Improve serialization / deserialization of SMs to make the state more easily accessible

### ISSUE THREAD

**Issue by mirgee** [hyperledger/aries-framework-rs#143]
... without duplication of the state information.

**Comment by Patrik-Stas** [comment]
For sake of public transparency now we are under Hyperledger, let me elaborate as per our discussion.

Currently the aries protocol state machines are represented by various nested structs. Whenever these structures are serialized (calls such as connection.serialize(), disclosedProof.serialize(), ..etc.), default serialization algorithm provided by serde_json crate is used. 
This can make it difficult to change internal implementation of state machines without breaking serialization format. It also makes some information hard to access in serialized form. It can be expected as common pattern to have these serialized protocol objects stored in some storage and wanting to peform some sort of search on them, for example based on state machine status, or connection DDO. 

Therefore we should implement custom serialization for these Aries objects, such that in its serialized form, it
- should be simple enough to perform these common operations
- should be "stable" across different states of the state machine. For example, for serialized connection, there should be a stable xpath to access DDO of the counterparty (maybe something like `data.counterparty.current_ddo`)

**Comment by bobozaur** [comment]
Is this issue still valid today? We discussed serialization formats a bit before and I believe the conclusions were:
- We maintain backwards compatibility if the state machine structure changes ( by converting from the old format to the new one).
- We don't make guarantees of the serialization format itself, meaning that it can change "without notice". Hence consumers should not rely on the inner structure of the serialized data (like direct access into the JSON object).

**Comment by mirgee** [comment]
@bobozaur You are correct, I think this old issue is not relevant anymore as per the discussions you mention. Clients should not rely on the serialization format.
